shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs

```diff
@@ -336,6 +342,85 @@ fromCV (CV w vs) = foldr f (inject . Coin . fromIntegral $ w) vs
     f (CVPart policyId aname amount) acc =
       insert (+) policyId aname (fromIntegral amount) acc
 
+data CompactValue2 crypto
+  = CompactValueAdaOnly {-# UNPACK #-} !Word64
+  | CompactValueMultiAsset
+      {-# UNPACK #-} !Word64 -- ada
+      {-# UNPACK #-} !Word -- number of ma's
+      {-# UNPACK #-} !ShortByteString -- rep
+      {-
+       - The rep consists of three parts
+       - The first part is a sequence of triples (Word16, Word16, Word64),
+       - representing policy id index, asset name index, and quantity.
+       - The second part is a blob of all policy ids.
+       - The third part is the a blob of all asset names.
+       - The pid index and assetid index indicate where the string starts.
+       -
+       - We recover the pid by grabbing the sequence of 28 bytes after the index position.
+       - We recover the asset name by grabbing bytes until the index of the next asset name (or the end of the bytestring)
+       -
+       -
+       -}
+
+from :: forall crypto. (CC.Crypto crypto) => CompactValue2 crypto -> Value crypto
+from (CompactValueAdaOnly c) = Value (fromIntegral c) mempty
+from (CompactValueMultiAsset c numAssets rep) =
+  valueFromList (fromIntegral c) triples
+  where
+    getTripleAtIndex :: Int -> (Word16, Word16, Word64)
+    getTripleAtIndex i =
+      let start = i * 12
+       in ( readWord16 rep start,
+            readWord16 rep (start + 2),
+            readWord64 rep (start + 4)
```

Comment:
This will lead to reading word 64s at 4-byte alignment, not 8-byte alignment. If we can use aligned read/writes then we can use more direct impls of readWord64 etc, without having to use bit twiddling.

Reply:
Where are the more direct impls?

We could fix the alignment issue by bringing the Word64s to the front of the structure